Parquet Reader/writer for fixed-size list arrays

dexterduck opened the pull request:
# Which issue does this PR close?

Closes #4214 

# What changes are included in this PR?

- Parquet writer support for fixed-size lists
- Parquet reader support for fixed-size lists

Reader and writer both include unit tests.   
I also sanity-checked the reader and writer against `pyarrow` (although I wasn't able to verify writing lists with nulls against pyarrow due to [this](https://github.com/apache/arrow/issues/35692) outstanding issue).

# Are there any user-facing changes?

None other than the added type support

tustvold commented on parquet/src/arrow/arrow_writer/levels.rs:
```diff
@@ -371,6 +396,88 @@ impl LevelInfoBuilder {
+                let mut start_idx = None;
+                for idx in range.clone() {
+                    if nulls.is_valid(idx) {
+                        // Start a run of valid rows if not already inside of one
```
I don't think coercing the rows like this is correct, as the repetition of each row must be different

dexterduck replied:
Shouldn't the repetition level inside each row be handled by the child elements? My implementation seems to match the same logic used by `write_struct`, but maybe I'm missing something?

tustvold replied:
You are quite correct, it has been a while since I've had to touch this code :smile:

tustvold commented on parquet/src/arrow/arrow_writer/levels.rs:
```diff
@@ -1397,4 +1504,160 @@ mod tests {
+
+        let expected_level = LevelInfo {
+            def_levels: Some(vec![0, 0, 3, 3]),
+            rep_levels: Some(vec![0, 0, 0, 1]),
```
```suggestion
            rep_levels: Some(vec![0, 1, 0, 1]),
```
As written it is not correct

dexterduck replied:
I believe that the original is correct since the levels are generated by `builder.write(&a, 1..4)`.

So the levels should correspond to the 1st, 2nd, and 3rd rows:
- def `0`, rep `0` (row 1, null)
- def `0`, rep `0` (row 2, null)
- def `3`, rep `0` (row 3, child 0)
- def `3`, rep `1` (row 3, child 1)

Is that not right?

tustvold reviewed (commented):
Thank you for this, I've left some comments. I think the level encoding is currently not quite correct, as it doesn't correctly delimit the row starts. 

I think the test coverage of this is pretty good, but I think the following would highlight where the current implementation is not quite right (or show that it is and I am mistaken :smile: ):

* A test reading a RecordBatch containing both a FixedSizeList column and a primitive column, this will test the record delimiting logic works correctly - as written I think it will error with length mismatch
* A test reading a parquet file with [skip_arrow_metadata](https://docs.rs/parquet/40.0.0/parquet/arrow/arrow_reader/struct.ArrowReaderOptions.html#method.with_skip_arrow_metadata), so it will infer the list as a regular List - this will again test that the level information is actually correct

dexterduck commented on parquet/src/arrow/arrow_writer/levels.rs:
```diff
@@ -371,6 +396,88 @@ impl LevelInfoBuilder {
+                    // Mark the start of each list in the child array
+                    for i in 0..num_items {
+                        let idx = start + i * fixed_size;
+                        rep_levels[idx] = ctx.rep_level - 1;
```
Correct me if I'm wrong, but my understanding is that the child elements' write methods should handle writing the leaf repetition levels So the fixed-size list writer should only need to set the repetition level to encode the start of each row.

tustvold replied:
Aah, I misread this the first time. I believe the logic isn't quite correct where you have a repeated child within the FixedSizeList. 

In particular it is making the erroneous assumption that there will be `fixed_size` leaves.

I wonder if we could make `write_list` have the signature

```
fn write_list<I, O: OffsetSizeTrait>(
        &mut self,
        ranges: I,
        nulls: Option<&NullBuffer>,
        values: &dyn Array,
    ) where I: IntoIterator<Item=(O, O)>
```

And use this to allow sharing the same logic :thinking:

dexterduck commented:
@tustvold thanks for the feedback! 

I implemented most of your corrections, and I added the two tests you suggested.
I also updated the test that writes a list of structs so that the structs contain 2 fields instead of 1, to try to catch any potential errors in the def/rep level encoding.

Following those changes, I believe the rep/def level writer logic is correct as-is, but it's possible I'm missing something.

tustvold commented on parquet/src/arrow/array_reader/fixed_size_list_array.rs:
```diff
@@ -0,0 +1,417 @@
+                        child_data_builder.extend_nulls(self.fixed_size);
+
+                        if let Some(validity) = validity.as_mut() {
+                            // Valid if empty list
```
Aah this is a good point, a zero-sized fixed size list is certainly a peculiar construction, but not technically invalid I don't think

tustvold commented on parquet/src/arrow/array_reader/fixed_size_list_array.rs:
```diff
@@ -0,0 +1,607 @@
+                    // - the list's fixed size (valid entries)
+                    // - zero (null entries, start of array)
+                    // Any other length indicates invalid data
+                    if row_len != 0 && row_len != self.fixed_size {
```
```suggestion
                    if start_idx.is_some() && row_len != self.fixed_size {
```

tustvold commented on parquet/src/arrow/array_reader/fixed_size_list_array.rs:
```diff
@@ -0,0 +1,607 @@
+            }
+            None => child_data_builder.freeze(),
+        };
+        assert_eq!(list_len * self.fixed_size, child_data.len());
```
I think this should return an error instead of panicking, as I think it can be encountered if the final row is the wrong length

tustvold reviewed (approved):
Thank you for this, I think this is mostly good to go. I don't think it correctly encodes levels for FixedSizeList with repeated children, but I'm also happy for that to be fixed as a follow on PR if you would prefer. Stellar work on this :muscle:

dexterduck commented:
@tustvold I've finally understood the encoding logic error you were pointing out 😅. 
I added tests for both reader and writer to verify correct encoding/decoding of fixed-length lists containing variable-length lists, and the incorrect behavior on the write side should now be fixed.

tustvold reviewed (approved):
Looks good to me, thank you
